shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java

```diff
@@ -154,18 +189,54 @@ public void setMessageHandler(
       @Nullable TaskQueue taskQueue) {
     if (handler == null) {
       Log.v(TAG, "Removing handler for channel '" + channel + "'");
-      messageHandlers.remove(channel);
-    } else {
-      DartMessengerTaskQueue dartMessengerTaskQueue = null;
-      if (taskQueue != null) {
-        dartMessengerTaskQueue = createdTaskQueues.get(taskQueue);
-        if (dartMessengerTaskQueue == null) {
-          throw new IllegalArgumentException(
-              "Unrecognized TaskQueue, use BinaryMessenger to create your TaskQueue (ex makeBackgroundTaskQueue).");
-        }
+      synchronized (handlersLock) {
+        messageHandlers.remove(channel);
       }
-      Log.v(TAG, "Setting handler for channel '" + channel + "'");
+      return;
+    }
+    DartMessengerTaskQueue dartMessengerTaskQueue = null;
+    if (taskQueue != null) {
+      dartMessengerTaskQueue = createdTaskQueues.get(taskQueue);
+      if (dartMessengerTaskQueue == null) {
+        throw new IllegalArgumentException(
+            "Unrecognized TaskQueue, use BinaryMessenger to create your TaskQueue (ex makeBackgroundTaskQueue).");
+      }
+    }
+    Log.v(TAG, "Setting handler for channel '" + channel + "'");
+
+    List<BufferedMessageInfo> list;
+    synchronized (handlersLock) {
       messageHandlers.put(channel, new HandlerInfo(handler, dartMessengerTaskQueue));
+      if (!bufferedMessages.containsKey(channel)) {
+        return;
+      }
+      list = bufferedMessages.remove(channel);
+    }
+    for (BufferedMessageInfo info : list) {
+      dispatchMessageToQueue(
+          channel, messageHandlers.get(channel), info.message, info.replyId, info.messageData);
+    }
+  }
+
+  @Override
+  public void enableBufferingIncomingMessages() {
+    enableBufferingIncomingMessages.set(true);
+  }
+
+  @Override
+  public void disableBufferingIncomingMessages() {
+    if (!enableBufferingIncomingMessages.getAndSet(false)) {
+      return;
+    }
+    Map<String, List<BufferedMessageInfo>> pendingMessagesCopy = new HashMap<>();
+    synchronized (handlersLock) {
+      pendingMessagesCopy = new HashMap(bufferedMessages);
```

Comment:
This copy can be avoided by setting `pendingMessages` to the current value of `bufferedMessages` and then setting `bufferedMessages` to a newly allocated map.

This is safe because `bufferedMessages` is only accessed while holding the `handlersLock`.